src/thread.rs

```diff
@@ -0,0 +1,583 @@
+        self.active_thread = id;
+        assert!(self.active_thread.index() < self.threads.len());
+        active_thread_id
+    }
+
+    /// Get the id of the currently active thread.
+    fn get_active_thread_id(&self) -> ThreadId {
+        self.active_thread
+    }
+
+    /// Get the total number of threads that were ever spawn by this program.
+    fn get_total_thread_count(&self) -> usize {
+        self.threads.len()
+    }
+
+    /// Has the given thread terminated?
+    fn has_terminated(&self, thread_id: ThreadId) -> bool {
+        self.threads[thread_id].state == ThreadState::Terminated
+    }
+
+    /// Get the borrow of the currently active thread.
+    fn active_thread_mut(&mut self) -> &mut Thread<'mir, 'tcx> {
+        &mut self.threads[self.active_thread]
+    }
+
+    /// Mark the thread as detached, which means that no other thread will try
+    /// to join it and the thread is responsible for cleaning up.
+    fn detach_thread(&mut self, id: ThreadId) -> InterpResult<'tcx> {
+        if self.threads[id].join_status != ThreadJoinStatus::Joinable {
+            throw_ub_format!("trying to detach thread that was already detached or joined");
+        }
+        self.threads[id].join_status = ThreadJoinStatus::Detached;
+        Ok(())
+    }
+
+    /// Mark that the active thread tries to join the thread with `joined_thread_id`.
+    fn join_thread(&mut self, joined_thread_id: ThreadId) -> InterpResult<'tcx> {
+        if self.threads[joined_thread_id].join_status != ThreadJoinStatus::Joinable {
+            throw_ub_format!("trying to join a detached or already joined thread");
+        }
+        if joined_thread_id == self.active_thread {
+            throw_ub_format!("trying to join itself");
+        }
+        assert!(
+            self.threads
+                .iter()
+                .all(|thread| thread.state != ThreadState::BlockedOnJoin(joined_thread_id)),
+            "a joinable thread has threads waiting for its termination"
+        );
+        // Mark the joined thread as being joined so that we detect if other
+        // threads try to join it.
+        self.threads[joined_thread_id].join_status = ThreadJoinStatus::Joined;
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+        Ok(())
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Get the name of the active thread.
+    fn get_thread_name(&mut self) -> InterpResult<'tcx, Vec<u8>> {
```

Comment:
Why not make this return `&[u8]` to avoid cloning if the caller does not need full ownership?